course6/toyapp_thomasdhuys/drawlib/primitives.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <iostream>
+
+#include "vertex.h"
+
+namespace Draw {
+
+class RawPrimitive;
+using Primitive = std::unique_ptr<RawPrimitive>;
+
+class RawPrimitive {
+public:
+    virtual Primitive Copy() = 0;
+	friend std::ostream& operator<<(std::ostream& s, const RawPrimitive& p);
+	friend std::ostream& operator<<(std::ostream& s, const std::unique_ptr<RawPrimitive>& p);
```

Comment:
oh this is just pure evil. Users of the API should know that they are dealing with pointer, and thus they should be comfortable with applying operator ```*``` where appropriate:
```
std::cout << "My primitive: " << *p << std::endl;
```